library/rmc/stubs/Rust/vec/rmc_vec.rs

```diff
@@ -0,0 +1,1085 @@
+                return;
+            }
+
+            // Call drop for elements which are truncated
+            let remaining_len = self.len - len;
+            while self.len != len {
+                self.len -= 1;
+                drop_in_place(self.as_mut_ptr().offset(self.len as isize));
+            }
+        }
+    }
+
+    // Clears the vector, removing all values.
+    // This method has no effect on the allocated capacity of the vector
+    pub fn clear(&mut self) {
+        self.truncate(0);
+    }
+
+    pub fn swap_remove(&mut self, index: usize) -> T {
+        let len = self.len;
+        assert!(index < len);
+
+        unsafe {
+            let last = read(self.as_ptr().add(len - 1));
+            let hole = self.as_mut_ptr().add(index);
+            self.set_len(len - 1);
+            let prev_hole = read(hole);
+            *hole = last;
+            prev_hole
+        }
+    }
+
+    // According to the RSL:
+    // "Returns the number of elements the vector can hold without reallocating."
+    // The API consumer cannot rely on the precision of this function.
+    pub fn capacity(&self) -> usize {
+        self.buf.capacity()
+    }
+
+    // Splits the collection into two at the given index.
+    //
+    // Returns a newly allocated vector containing the elements in the range [at, len). After the
+    // call, the original vector will be left containing the elements [0, at) with its previous
+    // capacity unchanged.
+    pub fn split_off(&mut self, at: usize) -> Self
+    where
+        A: Clone,
+    {
+        assert!(at <= self.len);
+
+        let other_len = self.len - at;
+        let mut other = Vec::with_capacity_in(other_len, self.allocator().clone());
+
+        unsafe {
+            // Copy all the elements from "at" till the end of the Vector through
+            // a memcpy which is much cheaper than remove() and push()
+            libc::memcpy(
+                other.as_mut_ptr() as *mut libc::c_void,
+                self.as_ptr().offset(at as isize) as *mut libc::c_void,
+                other_len * mem::size_of::<T>(),
+            );
+
+            // Set length to point to end of array.
+            self.set_len(at);
+            other.set_len(other_len);
+        }
+
+        other
+    }
+
+    pub fn append(&mut self, other: &mut Vec<T, A>) {
+        // Reserve enough space to reduce the number of resizing operations
+        self.reserve(other.len());
+        let mut i: usize = 0;
+        let olen = other.len();
+        // Create a temporary drain vector and add elements in reverse order so
+        // that the next time we pop, they are in increasing order of their
+        // original index
+        let mut drain = Vec::with_capacity(olen);
+        while i < olen {
+            drain.push(other.pop().unwrap());
+            i += 1;
+        }
+        // Empty other
+        i = 0;
+        while i < olen {
+            self.push(drain.pop().unwrap());
+            i += 1;
+        }
```

Comment:
Why can't this whole thing be a `memmove` followed by reducing the length of the second array?